wagtail/wagtailusers/views/groups.py

```diff
@@ -26,133 +22,152 @@ def get_permission_panel_classes():
-            'groups': groups,
-            'is_searching': is_searching,
-            'query_string': q,
-            'ordering': ordering,
-        })
-    else:
-        return render(request, "wagtailusers/groups/index.html", {
-            'search_form': form,
-            'groups': groups,
-            'is_searching': is_searching,
-            'ordering': ordering,
-            'query_string': q,
-        })
-
-
-@permission_required('auth.add_group')
-def create(request):
-    group = Group()
-    if request.method == 'POST':
-        form = GroupForm(request.POST, instance=group)
-        permission_panels = [
-            cls(request.POST, instance=group)
+class PermissionPanelFormsMixin(object):
+    def get_permission_panel_form_kwargs(self, cls):
+        kwargs = {}
+
+        if self.request.method in ('POST', 'PUT'):
+            kwargs.update({
+                'data': self.request.POST,
+                'files': self.request.FILES,
+            })
+
+        if hasattr(self, 'object'):
+            kwargs.update({'instance': self.object})
+
+        return kwargs
+
+    def get_permission_panel_forms(self):
+        return [
+            cls(**self.get_permission_panel_form_kwargs(cls))
             for cls in get_permission_panel_classes()
         ]
+
+    def get_context_data(self, **kwargs):
+        if 'permission_panels' not in kwargs:
+            kwargs['permission_panels'] = self.get_permission_panel_forms()
+
+        return super(PermissionPanelFormsMixin, self).get_context_data(**kwargs)
+
+
+class SearchableListMixin(object):
+    search_box_placeholder = _("Search")
+    search_fields = []
+
+    def get_search_form(self):
+        if 'q' in self.request.GET:
+            return SearchForm(self.request.GET, placeholder=self.search_box_placeholder)
+        else:
+            return SearchForm(placeholder=self.search_box_placeholder)
+
+    def get_queryset(self):
+        queryset = super(SearchableListMixin, self).get_queryset()
+        search_form = self.get_search_form()
+
+        if search_form.is_valid():
+            q = search_form.cleaned_data['q']
+
+            filters = {
+                field + '__icontains': q
+                for field in self.search_fields
+            }
+
+            queryset = queryset.filter(**filters)
+
+        return queryset
+
+    def get_context_data(self, **kwargs):
+        if 'search_form' not in kwargs:
+            kwargs['search_form'] = self.get_search_form()
+            kwargs['is_searching'] = 'q' in self.request.GET
+
+        return super(SearchableListMixin, self).get_context_data(**kwargs)
+
+
+class IndexView(SearchableListMixin, generic.IndexView):
+    page_title = _("Groups")
+    add_item_label = _("Add a group")
+    search_box_placeholder = _("Search groups")
+    search_fields = ['name']
+    context_object_name = 'groups'
+    paginate_by = 20
+    page_kwarg = 'p'
+    ordering = ['name']
```

Comment:
As part of reviewing this PR I noticed that because `wagtail.wagtailadmin.views.generic.IndexView` [overrides `get_queryset`](https://github.com/wagtail/wagtail/blame/master/wagtail/wagtailadmin/views/generic.py#L59), it doesn't properly support use of `ordering` in this way, unlike [`django.views.generic.list.BaseListView`](https://github.com/django/django/blob/master/django/views/generic/list.py#L42) as users might expect.

Was this omission intentional? If you remove that custom `get_queryset` (diff [here](https://gist.github.com/chosak/d0a5b682d3340ca92f1cecc46de6b916)), all mainline unit tests continue to pass, and it makes this PR easier to implement. Tagging @gasman as your name is on the blame for that code in Wagtail's `IndexView`; this change seems sensible to keep its behavior in line with Django's, but I wanted to see what you and @kaedroho thought. One concern: could this break ModelAdmin's ordering somehow?

Reply:
Also tagging @ababic for his input on the ModelAdmin ordering question.

Reply:
Hey @chosak / @tomdyson. The IndexView in ModelAdmin is still completely standalone I believe, so this change shouldn't effect it.

Reply:
Hi @chosak,
Until very recently (#3625) the `wagtail.wagtailadmin.views.generic` views were completely independent of Django's generic views - so if it turns out that we're now overriding Django's `get_queryset` with a less functional implementation, then that's almost certainly just an accident of the order things were written in, and we can indeed remove our version as redundant code. (It would definitely be good to check the two methods side by side to confirm that ours isn't doing anything subtly different, though, rather than trusting the unit tests to catch that :-) )

Reply:
I've removed these ``get_queryset`` methods in https://github.com/wagtail/wagtail/pull/3774/commits/251e41c1180a903888c06b58728f8721be6abe30. (They were also added to ``EditView`` and ``DeleteView`` but weren't being used by either of them)